### ISSUE TITLE
SSM and CSM environment

### ISSUE THREAD

**Issue by AFCMS** [minetest-toolkit/minetest-lsp-api#8]
**Why?**

Support of the SSM and CSM environment, with a `common` folder for the utilities that are the same in the two environments.

Basically to enable completion for SSM you will need to add both the path to the `ssm` folder and the `common` folder.

**The documentation should be mostly complete for ssm before doing so**

(dc7258d8e06d52e597c5f5984d9c7fa1daef47f4)

**Comment by lua-rocks** [comment]
I'm sorry for deleting this directory without asking. Maybe your extension won't work properly because of this, but let's make the full documentation first, and then you sort the files as needed.

**Comment by AFCMS** [comment]
Its fine, the extension is still far from complete anyways and I am a bit busy right now.

**Comment by lua-rocks** [comment]
Do I understand correctly that all definitions presented in [lua_api.txt] but absented in [client_lua_api.txt] must be separated in SSM directory?

[lua_api.txt]: https://github.com/minetest/minetest/blob/stable-5/doc/lua_api.txt
[client_lua_api.txt]: https://github.com/minetest/minetest/blob/stable-5/doc/client_lua_api.txt

**Comment by AFCMS** [comment]
Yes.

**Comment by lua-rocks** [comment]
I'm confused. I feel that many functions existed in `lua_api.txt` but missed in `client_lua_api.txt` should actually be there! Are helper functions `table.key_value_swap(t)` or `math.factorial(x)` available to the server only? This is very strange.

**Comment by AFCMS** [comment]
client_lua_api.txt is (sadly) outdated and doesn't always reflect changes made to the common APIs

To know if a function is common you can check if its defined in /builtin/common folder of minetest source code.

**Comment by lua-rocks** [comment]
@appgurueu Hi 🤗! Can you please confirm that client_api documentation is outdated in stable 5 branch or are we missing something?

**Comment by AFCMS** [comment]
Exemple:

https://github.com/minetest/minetest/blob/7f6b09dce82be5c2a881c075ca76c5ac8f84ad52/doc/client_lua_api.md?plain=1#L583-L603

https://github.com/minetest/minetest/blob/7f6b09dce82be5c2a881c075ca76c5ac8f84ad52/doc/lua_api.md?plain=1#L3477-L3681

**Comment by appgurueu** [comment]
Yes. See f.E. `math.factorial`. It might also be worth checking out https://github.com/minetest/minetest_docs for a few things, although mostly it only contains more in-depth docs than the Lua API.

Note that this doesn't mean that all SSM funcs are available to CSM, it just concerns the stuff defined in `builtin/common` as pointed out by AFCMS.
